direct/algorithms/optimization.py

```diff
@@ -0,0 +1,102 @@
+# coding=utf-8
+# Copyright (c) DIRECT Contributors
+
+"""General mathematical optimization techniques."""
+
+from typing import Callable, Optional
+
+import torch
+from torch import nn
+
+
+class Algorithm(nn.Module):
+    """Base class for implementing mathematical optimization algorithms."""
+
+    def __init__(self, max_iter: int = 30):
+        self.max_iter = max_iter
+        self.iter = 0
+        super().__init__()
+
+    def _update(self):
+        """Abstract method for updating the algorithm's parameters."""
+        raise NotImplementedError
```

Comment:
You can make a true abstract base class using `abc.ABC` and `@abc.abstractmethod`.